payara/Payara
# FISH-5687/FISH-1467 Integrate HotSwap Agent in Payara Platform

## Description
This is an experimental feature PR to integrate the Payara Platform with the HotSwap Agent.
HotSwap Agent is enabled by jvm-option `-XX:HotswapAgent` where VM name `Dynamic Code Evolution` is added to the option to skip this option for JDK with Hotspot vm.


## Important Info
### Todo

1. Currently, the HotSwap agent instrumenting all classloaders and loaded classes that need to be filtered and activate to the application classloader.
2. Modifying annotation or descriptor data is currently not supported and should be fixed under the following tickets:
      - FISH-1467 - Web Container
      -  FISH-5650 - EJB Container
      - FISH-5706 - JPA Container
      - FISH-5707 - Resources Container
      - FISH-5708 - CDI Container
      - FISH-5709 - Security Container
      - FISH-5710 - Connectors Container


## Testing
### Testing Performed
Manual tested on JDK11

### Step to test the feature
- To test this feature,  download the [Patched JDK](https://github.com/TravaOpenJDK/trava-jdk-11-dcevm/releases).
- Add the Payara Server to the Apache NetBeans IDE.
- Enable the Hot Deploy from the Properties of the registered server
- Start the Payara Server from command-line  (starting the Payara Server from IDE is not recommended currently as `VM name` support need to be added to the domain.xml parser in the IDE tools)
- Deploy the application from the IDE
- Change the Source code e.g Servlet
- You should see the similar following message in the server.log file.
```
HOTSWAP AGENT: 18:27:33.362 RELOAD (org.hotswap.agent.config.PluginManager) - Reloading classes [com.mycompany.mavenproject.NewServlet] (autoHotswap)]]
```
- And on refreshing the browser, changes should be reflected instantly.
